pcdsdevices/ccm.py

```diff
@@ -773,6 +774,198 @@ def inverse(self, real_pos: namedtuple) -> namedtuple:
+        ipms_prefix: str = None,
+        ipm_idx: typing.Optional[dict] = None,
+        hutch: typing.Optional[str] = None,
+        **kwargs
+    ):
+        # Put some effort into filling this automatically
+        # CCM exists only in two hutches
+        if hutch is not None:
+            self.hutch = hutch
+        elif 'XPP' in prefix:
+            self.hutch = 'XPP'
+        elif 'XCS' in prefix:
+            self.hutch = 'XCS'
+        else:
+            self.hutch = 'TST'
+
+        prefixes = {'ipms_prefix': ipms_prefix}
+        UCpt.collect_prefixes(self, prefixes)
+        super().__init__(prefix, **kwargs)
+
+        self.ipm_idx = ipm_idx or {'upstream': 0, 'downstream': 1}
+
+        # Optimization methods and arguments
+        # See function definitions for details on arguments
+        self.method = 'monotonic'
+        self.method_map = {
+            'monotonic' : self.monotonic_increase_search,
+        }
+        self.optimization_args = {}
+        self.optimization_args['monotonic'] = {
+            'offset': 5,
+            'n_average': 50,
+            'threshold_ratio': 2.5,
+            'stepsize': 0.5,
+            'max_step': 10
+        }
+
+    def monotonic_increase_search(
+        self,
+        setpoint: float,
+        offset: float,
+        n_average: int,
+        threshold_ratio: float,
+        stepsize: float,
+        max_steps: int
+    ):
+        """
+        Simple monotonic search for the CCM energy position:
+        Move CCM slightly under the setpoint and check if the downstream/upstream IPM
+        average is above a given threshold. If not, move the CCM energy up a bit and check
+        again.
+
+        Parameters
+        ----------
+        setpoint : float
+            Energy setpoint
+        offset: float
+            Energy offset. The initial CCM energy will be setpoint-offset.
+        n_average: int
+            Number of data points in the IPM average.
+        threshold_ratio: float
+            If mean(IPM downstream)/mean(IPM upstream)>threshold_ratio, the optimization
+            finishes.
+        stepsize: float
+            Energy increment after a failed check on the IPM ratio. Should be smaller
+            than the self-seed bandwidth for good results.
+        max_steps: int
+            If no good point is reached after max_step points, the optimization is
+            aborted.
+        """
+        self.self_seed.move(setpoint)
+        # move slightly below the setpoint
+        status = super().move(setpoint-offset, wait=True)
+        # wait for self-seed to settle
+        time.sleep(1)
+
+        for step in range(max_steps):
+            intensities = self.get_ipms_intensities(n_average)
+            ipm_ratio = self.ipm_ratio(intensities)
+            if ipm_ratio > self.ipm_ratio_threshold:
+                print('Optimization complete!\n')
+                break
+            status = super().mvr(stepsize, wait=True)
+        return status
```

Comment:
I think returning this status and using it in `bluesky` should work, though note that this will stall out the `asyncio` thread for the duration of the optimization which could be a problem if there were other background tasks. You don't need to prematurely do this though, it might not become a problem at all.

If it becomes a problem, you'd usually return the incomplete status immediately and do all the steps in a background thread and have the thread mark the status as finished at the appropriate time.